akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ConsistentHashingShardAllocationStrategy.scala

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (C) 2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.cluster.sharding
+
+import scala.collection.immutable
+import scala.concurrent.Future
+
+import akka.actor.ActorRef
+import akka.actor.ActorSystem
+import akka.actor.Address
+import akka.cluster.Cluster
+import akka.cluster.ClusterEvent.CurrentClusterState
+import akka.cluster.Member
+import akka.cluster.sharding.ShardCoordinator.ActorSystemDependentAllocationStrategy
+import akka.cluster.sharding.ShardRegion.ShardId
+import akka.cluster.sharding.internal.ClusterShardAllocationMixin
+import akka.cluster.sharding.internal.ClusterShardAllocationMixin.ShardSuitabilityOrdering
+import akka.event.Logging
+import akka.event.LoggingAdapter
+import akka.routing.ConsistentHash
+
+object ConsistentHashingShardAllocationStrategy {
+  private val emptyRebalanceResult = Future.successful(Set.empty[ShardId])
+}
+
+/**
+ * [[akka.cluster.sharding.ShardCoordinator.ShardAllocationStrategy]] that is using consistent
+ * hashing. This can be useful when shards with the same shard id for different entity types
+ * should be best effort collocated to the same nodes.
+ *
+ * When adding or removing nodes it will rebalance according to the new consistent hashing,
+ * but that means that only a few shards will be rebalanced and others remain on the same
+ * location.
+ *
+ * A good explanation of Consistent Hashing:
+ * https://tom-e-white.com/2007/11/consistent-hashing.html
+ *
+ */
+class ConsistentHashingShardAllocationStrategy(rebalanceLimit: Int)
```

Comment:
Note that it's fine to change to this allocation strategy in a rolling update. It's only used by the ShardCoordinator and when switching to new coordinator it will keep current allocations until rebalance and new allocations of shards.